### PR TITLE
add some tests for RegEx

### PR DESCRIPTION
**Pull request by swolchok**
Just some unit tests to catch regressions.

### CODE REVIEW

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -20,7 +20,7 @@ file(GLOB test_new_api_sources new-api/[a-z]*.cpp)
 list(APPEND test_sources ${test_new_api_sources})
 add_sources(${test_sources} ${test_headers})
 
-include_directories(${YAML_CPP_SOURCE_DIR}/test)
+include_directories(${YAML_CPP_SOURCE_DIR}/test ${YAML_CPP_SOURCE_DIR}/src)
```
Comment (jbeder) [inline test/CMakeLists.txt]: Is this change necessary?

File: test/regex_test.cpp [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+using YAML::RegEx;
+using YAML::Stream;
+
+namespace {
+TEST(RegExTest, Empty) {
+  RegEx empty;
+  EXPECT_TRUE(empty.Matches(std::string()));
+  EXPECT_EQ(0, empty.Match(std::string()));
+  for (int ii = Stream::eof() + 1; ii < 128; ++ii) {
```
Comment (jbeder) [inline test/regex_test.cpp]: 1) I'd prefer "i" to "ii"

2) Starting with Stream::eof() + 1 is probably not necessary. I'd hard-code the constant here. If you prefer, feel free to make a named constant at the top of this file (and if you do that, do one for the 128 as well.

File: test/regex_test.cpp [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    EXPECT_FALSE(empty.Matches(char(ii)));
```
Comment (jbeder) [inline test/regex_test.cpp]: Rather than casting to char, what do you think about making the loop variable a char? You won't be able to have < as the terminus of the loop, but maybe this way you could make, e.g., a constexpr vector<char> at the top of this file that consists of all of the chars that you're trying to test, and then just loop through that.

File: test/regex_test.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      } else {
+        EXPECT_FALSE(ex.Matches(char(jj)));
+        EXPECT_EQ(-1, ex.Match(char(jj)));
+      }
+    }
+  }
+}
+
+TEST(RegExTest, Range) {
+  for (int ii = Stream::eof() + 1; ii < 127; ++ii) {
```
Comment (jbeder) [inline test/regex_test.cpp]: Do you mean to ensure that ii < jj? If so, just make the second loop go from ii + 1 to 128.

File: test/regex_test.cpp [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+          EXPECT_FALSE(ex.Matches(char(kk)));
+          EXPECT_EQ(-1, ex.Match(char(kk)));
+        }
+      }
+    }
+  }
+}
+
+TEST(RegExTest, EmptyString) {
+  RegEx ex = RegEx(std::string());
```
Comment (jbeder) [inline test/regex_test.cpp]: "" rather than std::string()

File: test/regex_test.cpp [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  EXPECT_TRUE(ex.Matches(std::string()));
+  EXPECT_EQ(0, ex.Match(std::string()));
+
+  // Matches anything, unlike RegEx()!
+  EXPECT_TRUE(ex.Matches(std::string("hello")));
+  EXPECT_EQ(0, ex.Match(std::string("hello")));
+}
+
+TEST(RegExTest, SingleCharacterString) {
```
Comment (jbeder) [inline test/regex_test.cpp]: If you're testing this, what's the point in testing the SingleCharacter, given that you even had to add to the public API in order to test individual characters?

File: test/regex_test.cpp [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+#include "regex_yaml.h"
+#include "stream.h"
+
+using YAML::RegEx;
+using YAML::Stream;
+
+namespace {
+TEST(RegExTest, Empty) {
+  RegEx empty;
+  EXPECT_TRUE(empty.Matches(std::string()));
```
Comment (jbeder) [inline test/regex_test.cpp]: "" instead of std::string()

File: test/regex_test.cpp [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+  }
+}
+
+TEST(RegExTest, EmptyString) {
+  RegEx ex = RegEx(std::string());
+  EXPECT_TRUE(ex.Matches(std::string()));
+  EXPECT_EQ(0, ex.Match(std::string()));
+
+  // Matches anything, unlike RegEx()!
+  EXPECT_TRUE(ex.Matches(std::string("hello")));
```
Comment (jbeder) [inline test/regex_test.cpp]: And throughout; why not just the string literals?

**Comment by jbeder** [comment]
This is a great addition, thanks!

File: test/regex_test.cpp (hunk 4 above)
Comment (swolchok) [inline test/regex_test.cpp]: I intended to test j > i as well. I figured testing everything is easy and cheap.

File: test/regex_test.cpp (hunk 6 above)
Comment (swolchok) [inline test/regex_test.cpp]: I originally intended to make some changes to the RegEx implementation that would have made this make sense, but they didn't pan out and you're right that it doesn't make sense as is.

**Comment by swolchok** [comment]
Not sure if you deleted the comments or I just can't find them, but I see in my email that you asked why I call the `std::string` ctor explicitly. It's because there are templated versions of the ctor and Match that are a better match for `const char *` than the `std::string` versions are.

**Comment by jbeder** [comment]
Cherry-picked as 0a1352525aa4decefa6cc5e8ae9f0962175f5753.
